webpush/vapid.go

```
@@ -0,0 +1,124 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package webpush
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/json"
+	"math/big"
+	"net/url"
+	"time"
+)
+
+var (
+	vapidPrefix = []byte("eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NiJ9.")
+	dot         = []byte(".")
+)
+
+type jwtPrefix struct {
+}
+
+type jwtBody struct {
+	Aud string `json:"aud"`
+	Sub string `json:"sub,omitempty"`
+	Exp int64  `json:"exp"`
+}
+
+// Vapid represents a sender identity.
+type Vapid struct {
+	// The EC256 public key, base64 urlsafe, uncompressed. This value should be
+	// used in 'subscribe' requests and is included as p256ecdsa in
+	// the Crypto-Key header.
+	PublicKey string
+
+	// The private key used to sign tokens
+	pkey *ecdsa.PrivateKey
+
+	// Sub should be an email or URL, for identification
+	Sub string
+}
+
+// TODO: cache for tokens
+
+// Token creates a token with the specified endpoint, using configured Sub id
+// and a default expiration (1h).
+func (vapid *Vapid) Token(aud string) (res string) {
+	url, _ := url.Parse(aud)
```

Don't ignore the error